File: java/src/detectors/java-serversiderequestforgery/ServerSideRequestForgeryNoncompliant.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

// {fact rule=java-serversiderequestforgery@v1.0 defects=1}
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;

public class ServerSideRequestForgeryNoncompliant {
    // Noncompliant: Directly uses user-supplied URL without validation, potentially allowing SSRF attacks.
    public void nonCompliant(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        URL url = new URL(req.getParameter("url"));
        HttpURLConnection httpConnection = (HttpURLConnection) url.openConnection();
    }
}
// {/fact}

File: kotlin/src/detectors/server-side-request-forgery/server-side-request-forgery-non-compliant.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

// {fact rule=kotlin-server-side-request-forgery@v1.0 defects=1}
import io.ktor.server.engine.*
import io.ktor.server.netty.*
import io.ktor.server.routing.*
import io.ktor.server.response.*
import io.ktor.server.application.*
import java.net.URL

// Noncompliant: User input data passed to `respondText()`.
fun ssrf_noncompliant() {
    embeddedServer(Netty, port = 8080) {
        routing {
            post("/proxy/{url}") {
                val url = call.request.queryParameters["url"]
                if (url != null) {
                    val data = URL(url).readText()
                    call.respondText(data)
                }
            }
        }
    }.start(wait = true)
}
// {/fact}